### PR TITLE
Separate KPOps API from the CLI

### PR DESCRIPTION
**Pull request by raminqaf**

### CODE REVIEW

File: kpops/pipeline.py [hunk 1]
```diff
@@ -50,6 +48,36 @@ class Pipeline(BaseModel):
...
+    @staticmethod
+    def setup_handlers(config: KpopsConfig) -> ComponentHandlers:
+        schema_handler = SchemaHandler.load_schema_handler(config)
+        connector_handler = KafkaConnectHandler.from_kpops_config(config)
+        proxy_wrapper = ProxyWrapper(config.kafka_rest)
+        topic_handler = TopicHandler(proxy_wrapper)
+
+        return ComponentHandlers(schema_handler, connector_handler, topic_handler)
+
+    @computed_field(title="Step Names")
```
Comment (disrupted) [inline kpops/pipeline.py:76]: imo doesn't make sense to include this on model dump

File: kpops/pipeline.py [hunk 2]
```diff
@@ -50,6 +48,36 @@ class Pipeline(BaseModel):
 
     model_config = ConfigDict(arbitrary_types_allowed=True)
 
+    @classmethod
+    def create(
```
Comment (disrupted) [inline kpops/pipeline.py:52]: why not put this in api module? imo it doesn't belong to the Pipeline class

File: kpops/pipeline.py [hunk 3]
```diff
@@ -50,6 +48,36 @@ class Pipeline(BaseModel):
...
+        if kpops_config.components_module:
+            registry.find_components(kpops_config.components_module)
+        registry.find_components("kpops.components")
+
+        handlers = cls.setup_handlers(kpops_config)
+        parser = PipelineGenerator(kpops_config, registry, handlers)
+        return parser.load_yaml(pipeline_path, environment)
+
+    @staticmethod
+    def setup_handlers(config: KpopsConfig) -> ComponentHandlers:
```
Comment (disrupted) [inline kpops/pipeline.py:68]: since those are component handlers I wouldn't add it here

**Review by disrupted — CHANGES_REQUESTED** [review]
nice! 🤩

File: kpops/api/registry.py
Comment (disrupted) [inline kpops/api/registry.py:1]: how about `kpops.registry`?

File: kpops/api/api.py
Comment (disrupted) [inline kpops/api/api.py:1]: how about `kpops/api/__init__.py` to avoid the double api?

File: kpops/api/custom_formatter.py
Comment (disrupted) [inline kpops/api/custom_formatter.py:1]: how about `logs` or `logging`? we could move some more stuff to it (see other comment)

File: kpops/api/api.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+log = logging.getLogger("")
+LOG_DIVIDER = "#" * 100
+
+
+def log_action(action: str, pipeline_component: PipelineComponent):
+    log.info("\n")
+    log.info(LOG_DIVIDER)
+    log.info(f"{action} {pipeline_component.name}")
+    log.info(LOG_DIVIDER)
+    log.info("\n")
```
Comment (disrupted) [inline kpops/api/api.py:46]: we could move this to the logs module

File: kpops/cli/main.py [hunk 5]
```diff
@@ -1,43 +1,23 @@
 from __future__ import annotations
 
-import asyncio
-import logging
 from pathlib import Path
-from typing import TYPE_CHECKING, Optional
+from typing import Optional
 
 import dtyper
```
Comment (disrupted) [inline kpops/cli/main.py:6]: now that we have a proper `api` module we can remove the dependency on `dtyper`

File: kpops/api/api.py
Comment (raminqaf) [reply kpops/api/api.py:1]: after moving it into the __init__ file I get a -.-
```
E   ImportError: cannot import name 'KafkaConnectorConfig' from partially initialized module 'kpops.component_handlers.kafka_connect.model' (most likely due to a circular import) (/Users/ramingharib/Developer/bakdata/kpops/kpops/component_handlers/kafka_connect/model.py)

```

Comment (disrupted) [reply kpops/api/api.py:1]: that's weird. for me all the tests are passing after renaming it

File: kpops/api/registry.py
Comment (raminqaf) [reply kpops/api/registry.py:1]: Even moving this cuases 
```
E   ImportError: cannot import name 'KafkaConnectorConfig' from partially initialized module 'kpops.component_handlers.kafka_connect.model' (most likely due to a circular import) (/Users/ramingharib/Developer/bakdata/kpops/kpops/component_handlers/kafka_connect/model.py)
```

Comment (disrupted) [reply kpops/api/registry.py:1]: alright, then let's leave it as-is

File: kpops/cli/main.py [hunk 6]
```diff
@@ -126,99 +105,14 @@
...
-    if not path.exists():
-        path.mkdir(parents=False)
-    elif next(path.iterdir(), False):
-        log.warning("Please provide a path to an empty directory.")
-        return
-    init_project(path, config_include_opt)
+    kpops.init(path, config_include_opt=config_include_opt)
 
 
 @app.command(  # pyright: ignore[reportCallIssue] https://github.com/rec/dtyper/issues/8
```
Comment (disrupted) [inline kpops/cli/main.py:118]: those dtyper-specific Pyright pragmas can be removed now

**Review by disrupted — CHANGES_REQUESTED** [review]

File: kpops/api/__init__.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+def parse_steps(steps: str) -> set[str]:
+    return set(steps.split(","))
+
+
+def generate(
+    pipeline_path: Path,
+    dotenv: list[Path] | None = None,
+    config: Path = Path(),
+    steps: str | None = None,
```
Comment (disrupted) [inline kpops/api/__init__.py:39]: for the Python API function signature I would prefer directly using `set[str] | None`, `parse_steps` should be moved to the CLI

**Review by disrupted — CHANGES_REQUESTED** [review]

File: tests/cli/test_handlers.py
[the remark this answers is not in the document]
Comment (raminqaf) [reply tests/cli/test_handlers.py:1]: They can but I was not 100% this is a good idea. we have named them `test_generate` or `test_manifest` and have a mix of both inside and they are located in the `tests/pipeline`. I moved the `test_registry` to `tests/api` though

[the remark this answers is not in the document]
Comment (disrupted) [reply tests/cli/test_handlers.py:1]: all good, I mostly meant this and the other ones in `tests/cli`

**Review by disrupted — APPROVED** [review]
LGTM, great effort! 🙌🏻

**Pull request merged**